File: research/python-lab/src/edgechain_lab/mars/temporal.py
"""Temporal consistency scoring."""

from __future__ import annotations

from collections import defaultdict, deque
from typing import cast

import numpy as np
from numpy.typing import NDArray

from edgechain_lab.fl.metrics import cosine_similarity


class TemporalHistory:
    """Rolling per-site update history."""

    def __init__(self, max_history: int = 10) -> None:
        self._history: defaultdict[str, deque[np.ndarray]] = defaultdict(
            lambda: deque(maxlen=max_history)
        )

    def score_and_update(self, site_id: str, update: np.ndarray) -> float:
        """Score current update against historical direction and then store it."""

        history = self._history[site_id]
        if len(history) < 2:
            history.append(update)
            return 0.7

        ema = _ema(list(history))
        score = float(np.clip(cosine_similarity(update, ema), 0.0, 1.0))
        history.append(update)
        return score


def _ema(history: list[np.ndarray], decay: float = 0.8) -> np.ndarray:
    weights = np.array([decay**idx for idx in range(len(history))], dtype=float)
    weights = weights / weights.sum()
    ordered = list(reversed(history))
    result = np.sum(
        [weight * update for weight, update in zip(weights, ordered, strict=True)],
        axis=0,
    )
    return cast(NDArray[np.float64], result)

File: packages/mars/src/temporal.ts
import { clamp, cosineSimilarity } from './math';
import type { TemporalHistory as TemporalHistoryContract } from './types';

export class TemporalHistory implements TemporalHistoryContract {
  private readonly maxHistory: number;

  private readonly history = new Map<string, number[][]>();

  constructor(maxHistory = 10) {
    this.maxHistory = maxHistory;
  }

  scoreAndUpdate(siteId: string, update: number[]): number {
    const history = this.history.get(siteId) ?? [];
    if (history.length < 2) {
      this.append(siteId, history, update);
      return 0.7;
    }

    const score = clamp(cosineSimilarity(update, exponentialMovingAverage(history)));
    this.append(siteId, history, update);
    return score;
  }

  private append(siteId: string, history: number[][], update: number[]): void {
    const next = [...history, update].slice(-this.maxHistory);
    this.history.set(siteId, next);
  }
}

function exponentialMovingAverage(history: number[][], decay = 0.8): number[] {
  const ordered = [...history].reverse();
  const rawWeights = ordered.map((_, index) => decay ** index);
  const totalWeight = rawWeights.reduce((sum, weight) => sum + weight, 0);
  const weights = rawWeights.map((weight) => weight / totalWeight);
  const dimension = ordered[0]?.length ?? 0;

  return Array.from({ length: dimension }, (_, dimensionIndex) =>
    ordered.reduce((sum, update, updateIndex) => sum + weights[updateIndex] * (update[dimensionIndex] ?? 0), 0),
  );
}
